template/funcs.go

```diff
@@ -456,7 +464,192 @@ func pkiCertFunc(b *Brain, used, missing *dep.Set, destPath string) func(...stri
+		if len(s) == 0 {
+			return nil, nil
+		}
+
+		keyType := "rsa"
+		keyBits := 2048
+
+		var privateKey any
+		var rawKey string
+		var useCSRCommonName, useCSRSans bool
+
+		path, rest := s[0], s[1:]
+		data := make(map[string]interface{})
+		for _, str := range rest {
+			if len(str) == 0 {
+				continue
+			}
+			parts := strings.SplitN(str, "=", 2)
+			if len(parts) != 2 {
+				return nil, fmt.Errorf("not k=v pair %q", str)
+			}
+
+			k, v := strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])
+			// since we are generating the private key on our end we should not send
+			// key_type and key_bits to Vault
+			// use_csr_common_name and use_csr_sans here are meant to mirror the settings on the
+			// vault role side, so we can configure on our end accordingly
+			if k != "key_type" && k != "key_bits" && k != "use_csr_common_name" && k != "use_csr_sans" {
+				data[k] = v
+			}
+			// if we passed a key_type and the value is either rsa, ec or ed25519 we override the default value
+			if k == "key_type" && (v == "rsa" || v == "ed25519" || v == "ec") {
+				keyType = v
+			}
+			// if we passed key_bits we override the default value
+			if k == "key_bits" {
+				keyBit, err := strconv.Atoi(v)
+				if err != nil {
+					return nil, err
+				}
+				keyBits = keyBit
+			}
+			// check if we passed use_csr_common_name for later usage
+			if k == "use_csr_common_name" {
+				useCSRCommonName = true
+			}
+			// check if we passed use_csr_sans for later usage
+			if k == "use_csr_sans" {
+				useCSRSans = true
+			}
+		}
+
+		var csrTemplate x509.CertificateRequest
+
+		// if we passed use_csr_common_name, serverside will expect the commonname from the csr
+		// so besides adding that param to the csr template, we also remove it from the map and pass it later
+		// to vault; this way, we spare a warning from the server side
+		if useCSRCommonName {
+			commonName, ok := data["common_name"]
+			if ok {
+				csrTemplate.Subject.CommonName = commonName.(string)
+			}
+			delete(data, "common_name")
+		}
+		// if we passed use_csr_sans, that means serverside will expect the subject alternate names from the csr
+		// so besides adding that param to the csr template, we also remove it from the map we pass later
+		if useCSRSans {
+			subjectAltNames, ok := data["uri_sans"]
+			if ok {
+				csrTemplate.DNSNames = strings.Split(subjectAltNames.(string), ",")
+			}
+			subjectAltIPs, ok := data["ip_sans"]
+			if ok {
+				for _, ip := range strings.Split(subjectAltIPs.(string), ",") {
+					parsedIP := net.ParseIP(ip)
+					csrTemplate.IPAddresses = append(csrTemplate.IPAddresses, parsedIP)
+				}
+			}
+			delete(data, "uri_sans")
+			delete(data, "ip_sans")
+		}
+
+		// generating private keys and also pem encode them for later usage
+		if keyType == "rsa" {
+			key, err := rsa.GenerateKey(rand.Reader, keyBits)
+			if err != nil {
+				return nil, err
+			}
+			privateKey = key
+			csrTemplate.SignatureAlgorithm = x509.SHA512WithRSA
```

Comment:
What about looking for a `signature_bits` parameter to set the signature algorithm? Please consider it, as well as `use_pss` parameter for the RSA case.

https://github.com/hashicorp/vault/blob/371ffc4bd47a70c09e6b5c59cdd659ec09b719fa/sdk/helper/certutil/helpers.go#L1297-L1309